Update workspace docs for alpha release (#2836)
PrairieLearn/PrairieLearn: docs/workspaces/index.md

# Workspaces (Alpha release: NOT suitable for student use)

Workspaces allow students to work in persistent remote containers via in-browser frontends such as VS Code and JupyterLab. The remote containers are configured by instructors to provide custom, uniform environments per question. Workspace questions are integrated with the standard PrairieLearn autograding pipeline.

## Supported browsers

* [x] Chrome is supported
* [x] Firefox is supported
* [x] Edge Chromium (version >= 79) is supported
* [ ] Edge Legacy (version < 79) is untested

> * [ ] Safari is NOT yet supported

Reviewer: ```suggestion
* [x] Safari is supported
```

# Workspaces (Alpha release: NOT suitable for student use)

Workspaces allow students to work in persistent remote containers via in-browser frontends such as VS Code and JupyterLab. The remote containers are configured by instructors to provide custom, uniform environments per question. Workspace questions are integrated with the standard PrairieLearn autograding pipeline.

## Supported browsers

* [x] Chrome is supported
* [x] Firefox is supported

> * [ ] Safari is NOT supported

Reviewer: ```suggestion
* [ ] Safari is NOT yet supported
```

## Setting up

### `info.json`

The question's `info.json` should contain a `workspaceOptions` dictionary:

> * `image`: Dockerhub image serving the IDE and containing the desired compilers, debuggers, etc.

Reviewer: ```suggestion
* `image`: Docker Hub image serving the IDE and containing the desired compilers, debuggers, etc.
```

> * `port`: port number used in the Docker image

Reviewer: ```suggestion
* `port`: port number used by the workspace app inside the Docker image
```

> * `home`: home directory in the Docker image

Reviewer: ```suggestion
* `home`: home directory inside the Docker image
```

> * `gradedFiles`: list of files or directories that will be copied for grading

Reviewer: ```suggestion
* `gradedFiles`: list of files or directories that will be copied out of the workspace container for grading
```

> * `args` (optional): command line arguments to pass to the Docker image

Reviewer: ```suggestion
* `args` (optional, default none): command line arguments to pass to the Docker image
```

> * `syncIgnore` (optional): list of files or directories that will be excluded from sync

Reviewer: ```suggestion
* `syncIgnore` (optional, default none): list of files or directories that will be excluded from sync
```

> * `urlRewrite` (optional): if true, the URL will be rewritten such that the workspace container will see all requests as originating from /

Reviewer: ```suggestion
* `urlRewrite` (optional, default true): if true, the URL will be rewritten such that the workspace container will see all requests as originating from /
```

> A full `info.json` file should look something like:

Reviewer: ```suggestion
A full `info.json` file for a workspace question should look something like:
```

```

In Linux, `cd` to your course directory and copy-paste the following command (same as the MacOS command but add the `--add-host` option):

```sh
docker run -it --rm -p 3000:3000 \
  -v "$PWD":/course \
  -v "$HOME/pl_ag_jobs:/jobs" \
  -e HOST_JOBS_DIR="$HOME/pl_ag_jobs" \
  -v /var/run/docker.sock:/var/run/docker.sock \
  --add-host=host.docker.internal:172.17.0.1 \ # this line is new vs MacOS
  prairielearn/prairielearn \
  /PrairieLearn/docker/start_workspace.sh
```

In Windows 10 (PowerShell), `cd` to your course directory and copy the following command **except with your own username** in `HOST_JOBS_DIR`:

```powershell
docker run -it --rm -p 3000:3000 `
  -v $HOME\pl_ag_jobs:/jobs `
  -e HOST_JOBS_DIR=/c/Users/Tim/pl_ag_jobs `
  -v /var/run/docker.sock:/var/run/docker.sock `
  prairielearn/prairielearn `
  /PrairieLearn/docker/start_workspace.sh
```

**Note the following about `HOST_JOBS_DIR` in PowerShell:**

    * Use Unix-style paths (i.e., use `/c/Users/Tim/pl_ag_jobs`, not `C:\Users\Tim\pl_ag_jobs`).
    * Use the full path rather than $HOME (i.e., use `/c/Users/Tim/pl_ag_jobs`, not `$HOME/pl_ag_jobs`).

### Windows errors and quirks

> #### `exec user process caused "no such file or directory"`

Reviewer: Is this section needed? It looks like it's only a problem with the external graders in externally graded questions and won't be an issue for workspaces.

This error occurs during grading as a result of an OS new-line incompatibility with the `entrypoint` script in the externally
graded question:

```sh
standard_init_linux.go:207: exec user process caused "no such file or directory"
```

One solution for this is to make a `.gitattributes` files in your PL repository with the line
`*.sh text eol=lf`. This tells the GitHub client to write the script files in native Linux
format instead of converting them for Windows (which breaks mapping them back into docker).
This mimics the [`.gitattributes` file in the main PrairieLearn repo](https://github.com/PrairieLearn/PrairieLearn/blob/master/.gitattributes).

> #### `invalid mode: /grade`

Reviewer: Is this section needed for workspaces? It looks like it will only be a problem for externally graded questions, not workspaces.

Author: Replaced the "Windows quirks" section with a note about ensuring `C:` is shared.
